File: State Photos/state_photos/svg_map.py
"""Build interactive map SVG with optional cover image patterns."""
from __future__ import annotations

import xml.etree.ElementTree as ET
from pathlib import Path

from .cover_spec import CoverSpec
from .framing import pattern_image_transform
from .states import ALL_CODES

SVG_NS = "http://www.w3.org/2000/svg"

ET.register_namespace("", SVG_NS)


def _local_tag(tag: str) -> str:
    if tag.startswith("{"):
        return tag.split("}", 1)[-1]
    return tag


def _state_code_from_class(class_attr: str | None) -> str | None:
    if not class_attr:
        return None
    codes = {c.lower() for c in ALL_CODES}
    for token in class_attr.split():
        t = token.lower()
        if len(t) == 2 and t.isalpha() and t in codes:
            return t.upper()
    return None


def build_map_svg(
    svg_path: Path, covers: dict[str, CoverSpec], *, path_prefix: str = ""
) -> str:
    """
    covers: state_code -> cover spec (image URL + framing for pattern <image> transform)
    path_prefix: mount path when served under the main site (e.g. /statephotos).
    """
    base = (path_prefix or "").strip().rstrip("/")
    state_base = f"{base}/state" if base else "/state"
    tree = ET.parse(svg_path)
    root = tree.getroot()

    defs = root.find(f"{{{SVG_NS}}}defs")
    if defs is None:
        defs = ET.SubElement(root, f"{{{SVG_NS}}}defs")

    existing_pattern_ids = {
        el.get("id")
        for el in defs.iter()
        if _local_tag(el.tag) == "pattern" and el.get("id")
    }

    for code, spec in covers.items():
        pid = f"photo-pat-{code}"
        if pid in existing_pattern_ids:
            continue
        pattern = ET.SubElement(defs, f"{{{SVG_NS}}}pattern")
        pattern.set("id", pid)
        pattern.set("patternContentUnits", "objectBoundingBox")
        pattern.set("width", "1")
        pattern.set("height", "1")
        img = ET.SubElement(pattern, f"{{{SVG_NS}}}image")
        img.set("href", spec.href)
        img.set("width", "1")
        img.set("height", "1")
        img.set("preserveAspectRatio", "xMidYMid slice")
        img.set("transform", pattern_image_transform(spec.focus_x, spec.focus_y, spec.zoom))

    for el in root.iter():
        tag = _local_tag(el.tag)
        if tag == "path":
            code = _state_code_from_class(el.get("class"))
            if not code:
                continue
            if code == "DC":
                el.set("style", "display:none;")
                continue
            el.set("id", f"state-path-{code}")
            el.set("data-state", code)
            onclick = f"window.location.href='{state_base}/{code}';"
            if code in covers:
                el.set("fill", f"url(#photo-pat-{code})")
                el.set("style", f"cursor:pointer;{el.get('style') or ''}")
            else:
                el.set("style", "cursor:pointer;")
            el.set("onclick", onclick)
        elif tag == "circle":
            code = _state_code_from_class(el.get("class"))
            if code != "DC":
                continue
            el.set("id", "state-path-DC")
            el.set("data-state", "DC")
            onclick = f"window.location.href='{state_base}/DC';"
            if "DC" in covers:
                el.set("fill", "url(#photo-pat-DC)")
            el.set("style", "cursor:pointer;")
            el.set("onclick", onclick)

    return ET.tostring(root, encoding="unicode")

File: lib/statePhotos/svgMap.ts
import { DOMParser, XMLSerializer } from "@xmldom/xmldom";

import { ALL_CODES } from "./states";
import { patternImageTransform } from "./framing";

const SVG_NS = "http://www.w3.org/2000/svg";

/** xmldom document (avoid clashing with DOM lib `Document` typings). */
type SvgDom = {
  documentElement: Element;
  getElementsByTagName: (name: string) => ArrayLike<Element>;
  createElementNS: (namespaceURI: string | null, qualifiedName: string) => Element;
};

export type CoverSpec = {
  href: string;
  focus_x: number;
  focus_y: number;
  zoom: number;
};

function codeFromClass(classAttr: string | null): string | null {
  if (!classAttr) return null;
  const codes = new Set(ALL_CODES.map((c) => c.toLowerCase()));
  for (const token of classAttr.split(/\s+/)) {
    const t = token.toLowerCase();
    if (t.length === 2 && /^[a-z]{2}$/.test(t) && codes.has(t)) {
      return t.toUpperCase();
    }
  }
  return null;
}

function localTag(tag: string): string {
  return tag.includes("}") ? (tag.split("}", 2)[1] ?? tag) : tag;
}

function ensureDefs(doc: SvgDom): Element {
  const root = doc.documentElement;
  const list = doc.getElementsByTagName("defs");
  if (list.length > 0) return list[0] as Element;
  const defs = doc.createElementNS(SVG_NS, "defs");
  root.insertBefore(defs, root.firstChild);
  return defs;
}

export function buildMapSvg(
  svgXml: string,
  covers: Record<string, CoverSpec>,
  pathPrefix: string
): string {
  const base = pathPrefix.replace(/\/$/, "");
  const stateBase = base ? `${base}/state` : "/state";

  const doc = new DOMParser().parseFromString(svgXml, "image/svg+xml") as unknown as SvgDom;
  const root = doc.documentElement;
  const defs = ensureDefs(doc);

  const existingPatternIds = new Set<string>();
  for (let i = 0; i < defs.childNodes.length; i++) {
    const ch = defs.childNodes.item(i);
    if (ch?.nodeType === 1 && localTag((ch as Element).tagName) === "pattern") {
      const id = (ch as Element).getAttribute("id");
      if (id) existingPatternIds.add(id);
    }
  }

  for (const [code, spec] of Object.entries(covers)) {
    const pid = `photo-pat-${code}`;
    if (existingPatternIds.has(pid)) continue;
    existingPatternIds.add(pid);
    const pattern = doc.createElementNS(SVG_NS, "pattern");
    pattern.setAttribute("id", pid);
    pattern.setAttribute("patternContentUnits", "objectBoundingBox");
    pattern.setAttribute("width", "1");
    pattern.setAttribute("height", "1");
    const img = doc.createElementNS(SVG_NS, "image");
    img.setAttribute("href", spec.href);
    img.setAttribute("width", "1");
    img.setAttribute("height", "1");
    img.setAttribute("preserveAspectRatio", "xMidYMid slice");
    img.setAttribute("transform", patternImageTransform(spec.focus_x, spec.focus_y, spec.zoom));
    pattern.appendChild(img);
    defs.appendChild(pattern);
  }

  const paths = Array.from(doc.getElementsByTagName("path"));
  for (const el of paths) {
    const code = codeFromClass(el.getAttribute("class"));
    if (!code) continue;
    if (code === "DC") {
      el.setAttribute("style", "display:none;");
      continue;
    }
    el.setAttribute("id", `state-path-${code}`);
    el.setAttribute("data-state", code);
    const onclick = `window.location.href='${stateBase}/${code}';`;
    if (covers[code]) {
      el.setAttribute("fill", `url(#photo-pat-${code})`);
      el.setAttribute("style", `cursor:pointer;${el.getAttribute("style") ?? ""}`);
    } else {
      el.setAttribute("style", "cursor:pointer;");
    }
    el.setAttribute("onclick", onclick);
  }

  const circles = Array.from(doc.getElementsByTagName("circle"));
  for (const el of circles) {
    const code = codeFromClass(el.getAttribute("class"));
    if (code !== "DC") continue;
    el.setAttribute("id", "state-path-DC");
    el.setAttribute("data-state", "DC");
    const onclick = `window.location.href='${stateBase}/DC';`;
    if (covers.DC) {
      el.setAttribute("fill", "url(#photo-pat-DC)");
    }
    el.setAttribute("style", "cursor:pointer;");
    el.setAttribute("onclick", onclick);
  }

  return new XMLSerializer().serializeToString(root as never);
}
